src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/CrlCache.cs

```diff
@@ -0,0 +1,253 @@
+﻿using System;
+using System.IO;
+using System.Security.Cryptography.X509Certificates;
+using Microsoft.Win32.SafeHandles;
+
+namespace Internal.Cryptography.Pal
+{
+    internal static class CrlCache
+    {
+        public static void AddCrlForCertificate(
+            X509Certificate2 cert,
+            SafeX509StoreHandle store,
+            X509RevocationMode revocationMode,
+            DateTime verificationTime,
+            ref TimeSpan remainingDownloadTime)
+        {
+            // In Offline mode, accept any cached CRL we have.
+            // "CRL is Expired" is a better match for Offline than "Could not find CRL"
+            if (revocationMode != X509RevocationMode.Online)
+            {
+                verificationTime = DateTime.MinValue;
+            }
+
+            if (AddCachedCrl(cert, store, verificationTime))
+            {
+                return;
+            }
+
+            // Don't do any work if we're over limit or prohibited from fetching new CRLs
+            if (remainingDownloadTime <= TimeSpan.Zero ||
+                revocationMode != X509RevocationMode.Online)
+            {
+                return;
+            }
+
+            DownloadAndAddCrl(cert, store, ref remainingDownloadTime);
+        }
+
+        private static bool AddCachedCrl(X509Certificate2 cert, SafeX509StoreHandle store, DateTime verificationTime)
+        {
+            string crlFile = GetCachedCrlPath(cert);
+
+            using (SafeBioHandle bio = Interop.libcrypto.BIO_new_file(crlFile, "rb"))
+            {
+                if (bio.IsInvalid)
+                {
+                    return false;
+                }
+
+                // X509_STORE_add_crl will increase the refcount on the CRL object, so we should still
+                // dispose our copy.
+                using (SafeX509CrlHandle crl = Interop.libcrypto.PEM_read_bio_X509_CRL(bio))
+                {
+                    if (crl.IsInvalid)
+                    {
+                        return false;
+                    }
+
+                    // If crl.LastUpdate is in the past, downloading a new version isn't really going
+                    // to help, since we can't rewind the Internet. So this is just going to fail, but
+                    // at least it can fail without using the network.
+                    //
+                    // If crl.NextUpdate is in the past, try downloading a newer version.
+                    DateTime nextUpdate = OpenSslX509CertificateReader.ExtractValidityDateTime(
+                        Interop.Crypto.GetX509CrlNextUpdate(crl));
+
+                    if (nextUpdate < verificationTime)
```

Comment:
I was about to say "no, we take care of that", but that's when sending the date into the chain for full verification.

if verificationTime is DateTimeKind.Utc there might be an offset conversion that doesn't happen... I'll play around with this and determine the answer; and fix/document it when so determined.